File: src/main/java/com/github/zoewithabang/service/UserService.java
package com.github.zoewithabang.service;

import com.github.zoewithabang.dao.UserDao;
import com.github.zoewithabang.model.UserData;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.Properties;

public class UserService implements IService
{
    private Properties botProperties;
    private String database;
    private UserDao userDao;
    
    public UserService(Properties botProperties)
    {
        this.botProperties = botProperties;
        database = botProperties.getProperty("dbdatabase");
        userDao = new UserDao(botProperties);
    }
    
    public boolean userIsStored(String userId) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            return userDao.get(connection, userId) != null;
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on getting User for ID '{}'.", userId, e);
            throw e;
        }
    }
    
    public UserData getUser(String userId) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            return userDao.get(connection, userId);
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on getting User for ID '{}'.", userId, e);
            throw e;
        }
    }
    
    public UserData getUserWithMessages(String userId) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            return userDao.getWithMessages(connection, userId);
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on getting User and messages for ID '{}'.", userId, e);
            throw e;
        }
    }
    
    public UserData storeNewUser(String userId, boolean tracked, int permissionRank) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            UserData user = new UserData(userId, tracked, permissionRank);
            userDao.store(connection, user);
            return user;
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on storing new User for ID '{}'.", userId, e);
            throw e;
        }
    }
    
    public void updateUser(UserData user) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            userDao.update(connection, user);
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on updating User '{}'.", user, e);
            throw e;
        }
    }
    
    public void deleteUserWithId(String userId) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            UserData user = new UserData(userId, false, 0);
            userDao.delete(connection, user);
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on deleting User with ID '{}'.", userId, e);
            throw e;
        }
    }
    
    public void updateRankWithId(String userId, int rank) throws SQLException
    {
        try(Connection connection = userDao.getConnection(database))
        {
            UserData user = userDao.get(connection, userId);
            user.setPermissionRank(rank);
            userDao.update(connection, user);
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on updating rank of the User ID {} to {}.", userId, rank, e);
            throw e;
        }
    }
}

File: src/main/kotlin/com/github/zoewithabang/service/UserService.kt
package com.github.zoewithabang.service

import com.github.zoewithabang.dao.UserDao
import com.github.zoewithabang.model.UserData
import org.apache.logging.log4j.LogManager

import java.sql.Connection
import java.sql.SQLException
import java.util.Properties

class UserService(private val botProperties: Properties) {
    private val logger = LogManager.getLogger("UserService")
    private val database: String = botProperties.getProperty("dbdatabase")
    private val userDao: UserDao = UserDao(botProperties)

    @Throws(SQLException::class)
    fun userIsStored(userId: String): Boolean {
        try {
            userDao.getConnection(database).use { connection -> return userDao.get(connection, userId) != null }
        } catch (e: SQLException) {
            logger.error("SQLException on getting User for ID '{}'.", userId, e)
            throw e
        }
    }

    @Throws(SQLException::class)
    fun getUser(userId: String): UserData? {
        try {
            userDao.getConnection(database).use { connection -> return userDao.get(connection, userId) }
        } catch (e: SQLException) {
            logger.error("SQLException on getting User for ID '{}'.", userId, e)
            throw e
        }
    }

    @Throws(SQLException::class)
    fun getUserWithMessages(userId: String): UserData {
        try {
            userDao.getConnection(database).use { connection -> return userDao.getWithMessages(connection, userId) }
        } catch (e: SQLException) {
            logger.error("SQLException on getting User and messages for ID '{}'.", userId, e)
            throw e
        }
    }

    @Throws(SQLException::class)
    fun storeNewUser(userId: String, tracked: Boolean, permissionRank: Int): UserData {
        try {
            userDao.getConnection(database).use { connection ->
                val user = UserData(userId, tracked, permissionRank)
                userDao.store(connection, user)
                return user
            }
        } catch (e: SQLException) {
            logger.error("SQLException on storing new User for ID '{}'.", userId, e)
            throw e
        }
    }

    @Throws(SQLException::class)
    fun updateUser(user: UserData) {
        try {
            userDao.getConnection(database).use { connection -> userDao.update(connection, user) }
        } catch (e: SQLException) {
            logger.error("SQLException on updating User '{}'.", user, e)
            throw e
        }
    }

    @Throws(SQLException::class)
    fun deleteUserWithId(userId: String) {
        try {
            userDao.getConnection(database).use { connection ->
                val user = UserData(userId, false, 0)
                userDao.delete(connection, user)
            }
        } catch (e: SQLException) {
            logger.error("SQLException on deleting User with ID '{}'.", userId, e)

            throw e
        }
    }

    @Throws(SQLException::class)
    fun updateRankWithId(userId: String, rank: Int) {
        try {
            userDao.getConnection(database).use { connection ->
                val user = userDao[connection, userId]
                user!!.permissionRank = rank
                userDao.update(connection, user)
            }
        } catch (e: SQLException) {
            logger.error("SQLException on updating rank of the User ID {} to {}.", userId, rank, e)

            throw e
        }
    }
}
